File: C++/1.工厂模式/FactoryPatternDemo.cpp
/**
 * @file FactoryPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-10-07
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "ShapeFactory.h"

int main(int argc, char** argv) {
    std::shared_ptr<ShapeFactory> shape_factory = std::make_shared<ShapeFactory>();
    if (nullptr == shape_factory) return 0;

    std::shared_ptr<Shape> shape1 = shape_factory->GetShape("CIRCLE");
    if (nullptr != shape1) shape1->Draw();

    std::shared_ptr<Shape> shape2 = shape_factory->GetShape("RECTANGLE");
    if (nullptr != shape2) shape2->Draw();

    std::shared_ptr<Shape> shape3 = shape_factory->GetShape("SQUARE");
    if (nullptr != shape3) shape3->Draw();

    return 0;
}

File: Objective-C/1.FactoryPattern/FactoryPatternDemo.m
//
//  FactoryPatternDemo.m
//  1.FactoryPattern
//
//  Created by Layne on 2020/10/17.
//

#import "ShapeFactory.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        //获取 Circle 的对象，并调用它的 draw 方法
        id<ShapeProtocol> shape1 = [ShapeFactory getShape:@"CIRCLE"];

        //调用 Circle 的 draw 方法
        [shape1 draw];

        //获取 Rectangle 的对象，并调用它的 draw 方法
        id<ShapeProtocol> shape2 = [ShapeFactory getShape:@"RECTANGLE"];

        //调用 Rectangle 的 draw 方法
        [shape2 draw];

        //获取 Square 的对象，并调用它的 draw 方法
        id<ShapeProtocol> shape3 = [ShapeFactory getShape:@"SQUARE"];

        //调用 Square 的 draw 方法
        [shape3 draw];
    }
    return 0;
}
